server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -1052,9 +1063,13 @@ private MoveDecision decideMove(
                                 break;
                             }
                         } else if (bestDecision == Type.NOT_PREFERRED) {
-                            assert remainDecision.type() != Type.NOT_PREFERRED;
+                            assert remainDecision.type() != Type.NOT_PREFERRED || allocation.isSimulating() == false;
```

Comment:
This assertion is a bit confusing
```
assert remainDecision.type() != Type.NOT_PREFERRED || allocation.isSimulating() == false;
```

Do we need that `isSimulating() == false` clause? It looks like we should never get here when 
```
allocationDecision.type() == Type.NOT_PREFERRED && remainDecision.type() == Type.NOT_PREFERRED
```

Or have I missed something?

Reply:
You're right, that scenario is short circuited on L1049.
I forgot and got `!isSimulating`-happy after adding throttle below for the explain path.